app/models/workflow_step.rb

```diff
@@ -11,7 +11,8 @@ class WorkflowStep < ApplicationRecord
   validate  :workflow_exists, on: :create
   validate  :process_exists_for_workflow, on: :create
 
-  before_save :set_completed_at, if: :completed?
+  # Only set the completed_at field if status is completed and completed_at is unset
+  before_save :set_completed_at, if: :completed_and_not_already_completed?
```

Comment:
How about calling `before_save :maybe_set_completed`

Reply:
`before_save :set_completed_at, :if !already_completed?`  ?

Reply:
can you do that ?  do we even need `completed_and_not_already_completed?`

Reply:
or `:set_completed_at, :unless already_completed?`